### ISSUE TITLE
Implement WAND algorithm

### ISSUE THREAD

**Issue by fulmicoton** [quickwit-oss/tantivy#45]

**Comment by elshize** [comment]
@fulmicoton has this not been implemented? WAND is actually not that fast when compared to, e.g., (Variable) Block Max WAND but this one need an additional structure. In general, we found that Maxscore (which relies on the same structures as WAND) is more competitive: http://engineering.nyu.edu/~suel/papers/daat-ecir19.pdf

If that's something you still want to do, I'd be happy to look into it when I get some spare time.

**Comment by fulmicoton** [comment]
It would be awesome, but note it requires changing quite a bit of the architecture here :
the collector is not just this listener that gets document pushed to it. It needs to have some kind of feedback.

That is the main reason tantivy has not been implementing Block WAND yet.
If you still want to go into this rabbit hole, make sure to document your plans regarding the design so that they can be discussed openly.

**Comment by elshize** [comment]
I'll look at possible directions. Aside from the code design, there needs to be some additional index structure. Say, WAND relies on max scores per term, and block version is even more involved. You don't have maximum scores per term calculated already, do you? I didn't find anything but want to double check.

**Comment by fulmicoton** [comment]
I think we can jump to block wand directly and store the extra info at the block level.
For instance, replace the number of bits of bitpacking by the max term freq...

**Comment by elshize** [comment]
The downside of tying it into the encoding is that (a) the block size is fixed to whatever the encoding dictates, and (b) it seems rather problematic if ever you want to change the encoding, or provide an alternative.

In PISA, we simply have another "posting list" that is effectively a list of pairs `(last_doc, max_freq)`. So translating that into Tantivy, you'd have another `SegmentPostings`, and then you could pass it into something like
```rust
struct MaxBlockSegmentPosting {
    postings: SegmentPostings,
    max_blocks: SegmentPostings,
}
```
Then, either `Scorer` would also implement an additional method `max_score`, or more likely there is another `trait BlockMaxScorer` or something like this. The advantage of treating max blocks as if they were postings is (a) you can define whatever block size you want when indexing, and (b) it possibly makes implementation easier. The downside could be size of that structure, but then again, you would probably have it bit packed same as postings.

Does it make any sense?

**Comment by shikhar** [comment]
I believe this can be closed, as of tantivy 0.13? https://github.com/quickwit-oss/tantivy/blob/main/CHANGELOG.md#tantivy-0130

Followups:
* https://github.com/quickwit-oss/tantivy/issues/869
* https://github.com/quickwit-oss/tantivy/issues/1034

**Comment by fulmicoton** [comment]
Ah yes absolutely.
